Add page writer

sadikovi opened the pull request:
This PR adds: 
- `PageWriter` interface and serialised implementation.
- `CompressedPage` in addition to `Page`. Compressed page is a wrapper that allows us to store compressed buffer + uncompressed length. Page, when created, is always assumed to have uncompressed buffer. This is not the case for compressed page - internally we store compressed data.
- Changed `SerializedPageReader` to take `T: Read` instead of file source. This allows me to test a page roundtrip.
- Added tests.

sadikovi commented:
@sunchao I implemented page writer. Let me know if I need to add more tests, I feel like the currently added tests may not cover all of the cases. Thanks!

sunchao commented on src/file/writer.rs:
```diff
@@ -0,0 +1,447 @@
+
+  /// Serializes column chunk into Thrift.
+  #[inline]
+  fn serialize_column_chunk(&mut self, chunk: parquet::ColumnChunk) -> Result<()> {
```
Just for my understanding, this will only be called once per PageWriter instance, is that right?

sadikovi replied:
Yes, it will only be called when we finalise column writer. I know it is a bit confusing, it could be part of a column writer.

sadikovi commented on src/column/page.rs:
```diff
@@ -102,6 +103,65 @@ impl Page {
+  }
+
+  /// Returns underlying page with potentially compressed buffer.
+  pub fn get_compressed_page(&self) -> &Page {
```
Will do, thanks.

sadikovi commented on src/column/page.rs:
```diff
@@ -110,6 +170,39 @@ pub trait PageReader {
+  fn close(&mut self) -> Result<()>;
+
+  /// Returns dictionary page offset in bytes, if set.
+  #[inline]
```
Will do, thanks.

sadikovi commented on src/column/page.rs:
```diff
@@ -110,6 +170,39 @@ pub trait PageReader {
+  fn write_metadata(&mut self, metadata: &ColumnChunkMetaData) -> Result<()>;
+
+  /// Closes resources and flushes underlying sink.
+  fn close(&mut self) -> Result<()>;
```
Well, I thought about it, but this method is part of trait `PageWriter`, so I cannot make it `close(self)`.

sadikovi commented on src/file/writer.rs:
```diff
@@ -0,0 +1,447 @@
+    }
+  }
+
+  /// Serializes page header into Thrift.
```
Will do, thanks.

sadikovi commented on src/file/writer.rs:
```diff
@@ -0,0 +1,447 @@
+    match page_type {
+      PageType::DATA_PAGE | PageType::DATA_PAGE_V2 => {
+        if self.data_page_offset.is_none() {
+          self.data_page_offset = Some(start_pos);
```
Yes, every time we write a data page, we record the first offset. The same applies to dictionary page, but we will write only one of those. Normally, in column writer we would write either `DICTIONARY_PAGE, DATA_PAGE, ..., DATA_PAGE` or `DATA_PAGE, ..., DATA_PAGE`.

sunchao replied:
Got it. I misunderstood - I though the offset should be the start of the actual data, but it should be the start of the page header. Looks good now.

sadikovi commented:
@sunchao can you have a look again? I addressed your comments and added more docs describing each method.

I was thinking if `PageWriter` trait needs an improvement; it might be a bit difficult to reason about the API without column writer. Let me know what you think.

Thanks!

sadikovi commented:
Yes, you are right, it does. We should be able to refactor both when working on column chunk writer.

sunchao commented:
Merged. Thanks @sadikovi !

sadikovi commented:
Thanks @sunchao!
